### PR TITLE
Fix bracket URI serialization for LSP compatibility

### PR DESCRIPTION
**Pull request by WAKayser**
Fixes bracket URI serialization for LSP servers.

## Summary
- Add custom URL serialization that percent-encodes brackets ([, ])
- Update all URL fields across the codebase with custom serialization  
- Fix WorkspaceEdit URL map serialization to encode brackets
- Performance optimizations with fast path for URLs without brackets
- Comprehensive test coverage for edge cases

## Problem Solved
Resolves 'unexpected character at index X' errors when using files with brackets in names (like Deno Fresh framework's [slug].tsx dynamic routes).

## Testing
✅ Tested successfully in Zed GUI with Deno language server
✅ All existing tests pass
✅ Handles edge cases and performance scenarios

Fixes zed-industries/zed#21062

### CODE REVIEW

File: src/call_hierarchy.rs [hunk 1]
```diff
@@ -63,6 +63,7 @@ pub struct CallHierarchyItem {
     pub detail: Option<String>,
 
     /// The resource identifier of this item.
+    #[serde(with = "crate::lsp_url")]
```
Comment (SomeoneToIgnore) [inline src/call_hierarchy.rs]: I am quite confused: `Uri` comes out of `src/uri.rs` in the very same repo, and there are already `impl Serialize for Uri {` and `impl<'de> Deserialize<'de> for Uri {` there.

Why cannot those be changed instead, given that brackets in the path will be an issue across the whole spec?

If anything new is added, e.g. `CodeDescription` deserializer in https://github.com/zed-industries/lsp-types/pull/4/ (which is not covered by this PR at all??? Why?), everyone has to remember to add those magic annotations.

As a bonus, all custom types like workspace edits would not need any testing at all after this change: instead, a single test on URI [de]serialization is enough.

File: src/lib.rs [hunk 2]
```diff
@@ -994,7 +1093,9 @@ mod url_map {
             Some(ref changes) => {
                 let mut map = serializer.serialize_map(Some(changes.len()))?;
                 for (k, v) in changes {
-                    map.serialize_entry(k.as_str(), v)?;
+                    // Ensure brackets are encoded for LSP compatibility
```
Comment (SomeoneToIgnore) [inline src/lib.rs]: Let's not add useless comments that are covered by the code, verbatim.
Presumably, this change has to go away after the custom deserializer is removed.

File: tests/bracket_uri_test.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        let url_deserialized: Url = serde_json::from_str(&url_serialized).expect("Should deserialize URL");
+        
+        println!("Original: {}", url);
+        println!("Serialized: {}", url_serialized);
+        println!("Deserialized: {}", url_deserialized);
+        println!("---");
+        
+        assert_eq!(url, url_deserialized);
+    }
+}
```
Comment (SomeoneToIgnore) [inline tests/bracket_uri_test.rs]: Let's keep up with the rest of the files' style and add a trailing newline for each new file.

File: tests/bracket_uri_test.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+fn test_url_serialization_roundtrip() {
+    let test_cases = vec![
+        "file:///Users/test/[slug].tsx",
+        "file:///Users/test/blog/[id]/[slug].tsx",
+        "file:///Users/test/[[...slug]].tsx",
+        "file:///Users/test/[category]/[...slug].tsx",
+    ];
+    
+    for file_path in test_cases {
+        println!("Testing: {}", file_path);
```
Comment (SomeoneToIgnore) [inline tests/bracket_uri_test.rs]: Not sure any print is needed in such tests, let's remove them all.

File: src/lib.rs [hunk 5]
```diff
@@ -28,6 +28,99 @@ use serde::{
...
+            return serializer.serialize_str(url_str);
+        }
+        
+        // Slow path: encode brackets
+        let encoded_url = encode_brackets_in_url(url_str);
+        serializer.serialize_str(&encoded_url)
+    }
+
+    /// Deserialize URL from string
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Url, D::Error>
```
Comment (SomeoneToIgnore) [inline src/lib.rs]: Isn't this the same as the default deserializer?
Great to know that it's working well, but no need to repeat it: for custom case, `serde` has https://serde.rs/field-attrs.html#serialize_with but given the custom case is not supposed to be used here, it's the serialization alone that has to be changed.

**Review by SomeoneToIgnore — DISMISSED** [review]
Thank you for pushing this forward.

The direction is much better than adjusting Zed codebase, but the implementation has to be different.

File: src/call_hierarchy.rs (hunk 1 above)
Comment (WAKayser) [reply src/call_hierarchy.rs]: src/uri.rs only exists on the main branch and not on the branch that is being used by zed. "zed-main-no-url-changes".  That implementation is also based off of `fluent_uri` instead of `url`. I can update it to use a similar style.

Comment (SomeoneToIgnore) [reply src/call_hierarchy.rs]: Oh, thank you for the correction, I always miss that branch part in this repo.

We definitely need to cover all `Url` usages in this crate, one way or another, and now various `Option<Url>` are untouched it seems.
For wrapper `Foo<Url>` types, those `serde::deserialize` have to be implemented separately, every time, which means `HashMap<Url, DocumentDiagnosticReportKind>` and `Option<Url>` will require 2 more [de]serializers.

Not sure how feasible this is, as alternatives, I see (maybe, there are better ways?) 
* https://serde.rs/remote-derive.html -esque approach (not sure this will work for structs with existing impls)

* or another one, where one can define a `pub struct Url(pub other::Url)` in this crate, replace __all__ old Url types with the new one in the crate, then do 
```rs
impl Deref for Url {
    type Target = other::Url;
.....
}

impl From<other::Url> for Url {
....
}

+ proper serialize and deserialize impls on top
```

Zed uses this library only, so it's fine and most probably due to Deref, not that many things has to be changed there?

Comment (WAKayser) [reply src/call_hierarchy.rs]: Ive went ahead with just replacing all Url usages. The crate isnt so big that it is too much work to cover all uses.

File: tests/bracket_uri_test.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        "Location serialization should contain encoded brackets"
+    );
+    assert_eq!(
+        location.range, location_deserialized.range,
+        "Ranges should match"
+    );
+}
+
+#[test]
+fn test_workspace_edit_with_bracket_uris() {
```
Comment (SomeoneToIgnore) [inline tests/bracket_uri_test.rs]: * Then, this is not needed really?

`Uri` is now tested to [de]serialize properly, so whatever wrappers on top will automatically inherit this property, right?

Let's reduce the amount of extra tests  to be a bit more clear what's actually changed.

* Same confusion about `test_uri_with_brackets_serialization` and the `bracket_encoding_test.rs` — we do have a `test_uri_serialization_roundtrip` already that checks things, or am I missing some context again?

* Last but not least, we can move the remaining test(s) straight into `uri.rs`, cannot we?

File: src/uri.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
+use std::{hash::Hash, ops::Deref, str::FromStr};
+
+use serde::{de::Error, Deserialize, Serialize};
+
+/// Newtype struct around `url::Url` with serialization implementations that properly encode brackets for LSP compatibility.
+#[derive(Debug, Clone)]
+pub struct Uri(url::Url);
```
Comment (SomeoneToIgnore) [inline src/uri.rs:7]:
```suggestion
pub struct Uri(pub url::Url);
```

One concern I have is how Zed deals with this: if there's a need to construct the struct in Zed, there's no good way now but through the parsing.

File: src/uri.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // Encode brackets for LSP compatibility
+        let url_str = self.0.as_str();
+        if url_str.contains('[') || url_str.contains(']') {
+            // Fast bracket encoding - count brackets first for precise capacity
+            let bracket_count = url_str.chars().filter(|&c| c == '[' || c == ']').count();
+            let mut encoded = String::with_capacity(url_str.len() + bracket_count * 2);
```
Comment (SomeoneToIgnore) [inline src/uri.rs]:
```suggestion
            let mut encoded = String::with_capacity(url_str.len() + bracket_count * 3);
```

That has to be `3` given that we push `%5B` and another 3-sized item, right?

I think we can avoid that entirely and use iterators: `url_str.chars().map(|c| ...).collect::<String>()` to do that automatically, without any extra `let bracket_count` calculation and an `count brackets first for precise capacity` odd comment.

File: src/uri.rs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+impl PartialOrd for Uri {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl FromStr for Uri {
+    type Err = url::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```
Comment (SomeoneToIgnore) [inline src/uri.rs:56]: This and Deserialize are 2 separate code paths to create the `Uri`, so in whatever single test we'll have in the future, it might be good to check that both paths produce identical results.

**Review by SomeoneToIgnore — COMMENTED** [review]
Great, this is much more compelling, I have no "architecture" questions on this anymore, just the style ones mainly.

**Comment by WAKayser** [comment]
The current approach is not correct. It changes the interface with zed and requires updated imports. Ive also tested using deref but for patternmatching purposes there still need to be changes on the zed side.

File: src/uri.rs (hunk 8 above)
Comment (WAKayser) [reply src/uri.rs]: No it is two because you are replacing one char for two. But the iterator is nicer anyway.

**Comment by SomeoneToIgnore** [comment]
I have thought we have to now change all the places in Zed where `Url` is used — so, imports and field types mainly, which is unfortunate but somehow expected based on this PR's approach.

Yet, we do not have to use custom deserializers and adjust every single object inside Zed to handle [de]serialization differently as https://github.com/zed-industries/zed/pull/35746/files does?

So, the parsing logic stays in the library, and Zed uses whatever "url-ish" type where needed, no way around that.
That seems ok, but sure would be superb to see if anything else can be done.

Maybe, if could share the link on Zed's branch, it would be easier to reason about this?

**Comment by WAKayser** [comment]
On this commit I have made all the changes required in Zed to use the new type. Wasnt too much.

**Review by SomeoneToIgnore — APPROVED** [review]
Nice, thank you.

**Comment by imumesh18** [comment]
Hey @WAKayser sorry to jump on this. I was just curious on why not use https://github.com/servo/rust-url. Asking because this is currently getting used in lsprotocol(the official rust types lib for lsp from microsoft) and helix-lsp-types(helix's fork of lsp-types). It seems more broadly compliant with url standard.

**Comment by Sysix** [comment]
This only deals with percent encoding brackets.
What about other characters, which should percent encoded too, according to RFC 3986?
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uri

It was noticed in https://github.com/zed-industries/zed/issues/21062#issuecomment-3170975523 and https://github.com/zed-industries/zed/pull/35746#issuecomment-3170768420

**Comment by SomeoneToIgnore** [comment]
Sorry for an overlook, the original issue's title was about brackets too, and overall was quite bloated with comments.

Looks very logical to add more corner cases, given that Zed will stick with the fork for some time — luckily now we have an isolated module with the tests that should be super trivial to expand.

I'm fine for now until a next report of a kind, but happy to review whatever else related changes.

**Comment by WAKayser** [comment]
Yeah. It was an oversight to not encode the other characters. Not sure how often the other special characters happen in filenames.  No issue reports make it seem like it is not a big deal. 

Reading the spec makes it seem like encoding '%' is somewhat important. 

I think adding the rest should be easy enough.

**Comment by WAKayser** [comment]
Ok, i have just tested this with files containing the other characters that should be percent encoded via the spec. Seems like the rest of the files all work just fine with oxlint and eslint (which worked before). The original issue mentioned deno lsp specifically but also only for square brackets. https://github.com/WAKayser/zed-oxc-bug This now contains more files with special characters for testing.
